lib/jekyll/commands/serve/servlet.rb

```diff
@@ -21,10 +21,36 @@ def initialize(server, root, callbacks)
         # up with a different preference on which comes first.
 
         def search_file(req, res, basename)
-          # /file.* > /file/index.html > /file.html
+          # /file.* -> /file.html
           super || super(req, res, "#{basename}.html")
         end
 
+        def search_index_file(req, res)
+          # /file/index.html -> /file.html
+
+          # First, let's see if the our super method can figure it out.
+          # (i.e Check for index.html files in the res.filename directory)
+          file = super
+
+          unless file
+            # Ok, I guess that didn't work, I guess there's no basename/index.html
+            # Let's look for basename.html instead...
+
+            # Split the basename from the path (res.filename)
+            i = res.filename.rindex "\/"
+            basename = res.filename[i..-1]
+            res.filename = res.filename[0, i]
+
+            # Try and find a file named dirname.html in the parent directory.
+            file = search_file(req, res, basename + ".html")
```

Comment:
in the really rare chance that `basename` is not a `String`, this will fail. Instead interpolate:
```diff
- file = search_file(req, res, basename + ".html")
+ file = search_file(req, res, "#{basename}.html")
```